mqtt/mqtt-broker/tests/common.rs

```diff
@@ -284,11 +287,135 @@ where
     }
 }
 
+lazy_static! {
+    static ref DEFAULT_TIMEOUT: Duration = Duration::from_secs(5);
+}
+
+/// A simple wrapper around TcpStream + PacketCodec to send specific packets
+/// to a broker for more granular integration testing.
+#[derive(Debug)]
+pub struct PacketStream {
+    codec: Framed<TimeoutStream<TcpStream>, PacketCodec>,
+}
+
+impl PacketStream {
+    /// Creates a client and opens TCP connection to the server.
+    /// No MQTT packets are sent at this moment.
+    pub async fn open(server_addr: impl ToSocketAddrs) -> Self {
+        // broker may not be available immediately in the test,
+        // so we'll try to connect for some time.
+        let mut result = TcpStream::connect(&server_addr).await;
+        let start_time = Instant::now();
+        while let Err(_) = result {
+            if start_time.elapsed() > *DEFAULT_TIMEOUT {
+                break;
+            }
+            result = TcpStream::connect(&server_addr).await;
+        }
+
+        let tcp_stream = result.expect("unable to establish tcp connection");
+        let mut timeout = TimeoutStream::new(tcp_stream);
+        timeout.set_read_timeout(Some(*DEFAULT_TIMEOUT));
+        timeout.set_write_timeout(Some(*DEFAULT_TIMEOUT));
+
+        let codec = Framed::new(timeout, PacketCodec::default());
+
+        Self { codec }
+    }
+
+    /// Creates a client, opens TCP connection to the server,
+    /// and sends CONNECT packet.
+    pub async fn connect(
+        client_id: ClientId,
+        server_addr: impl ToSocketAddrs,
+        username: Option<String>,
+        password: Option<String>,
+    ) -> Self {
+        let mut client = Self::open(server_addr).await;
+        client
+            .send_connect(Connect {
+                username,
+                password,
+                client_id,
+                ..Default::default()
```

Comment:
what is the  `..` doing? I've never seen this syntax before

Reply:
https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax

Reply:
cool, good to know